### ISSUE TITLE
Masking out padded positions for attention

### ISSUE THREAD

**Issue by sofiaherrero** [ilivans/tf-rnn-attention#8]
Hi ilivans,

Great implementation! I was just wondering if the padded sequences don't have to be masked out somehow for the attention layer. I imagine that the logits in rnn_outputs have values up to max_len, so they include the padded positions (which the Bilstm has ignored) and then this is used for the calculations in attention. Do we need to set their attention scores to zero or something like that, or is the network able to learn that the padded positions should not be taken into account?

Thanks in advance.

**Comment by ilivans** [comment]
Hi @SofiaHerrero ,

Thank you for a good question. 
As far as I remember in TF the hidden states of RNN after seq_len position are filled with zeros, so this can be considered as some kind of low-level masking already. Strictly speaking the coefficients at the padded positions are still can take any values (because of the bias term in the formulas of attention), but this approach seems reasonable because of its simplicity and effectiveness. Moreover zeroed states after seq_len don't contribute to the final linear combination of all states with their coefficients, no matter how large the coefficients at those positions are. I don't actually know a better way of handling this problem, and would be happy to hear any suggestion.
Practically attention coefficients at the padded positions are very close to zero. Also if you notice that these coefficients values are far from zero, then probably it's a good indicator of underfitting or some bug.

**Comment by sofiaherrero** [comment]
Hi ilivans,

Thank you for your answer.  For some reason, I thought that the padded positions were calculated and then ignored but indeed I just checked that they are set to zero for the output. This will make the attention score zero which is what I was thinking on doing in case I needed to mask the padding positions for attention. So all solved. Thanks!
